File: src/main/java/nl/martijndwars/webpush/PushService.java
package nl.martijndwars.webpush;

import com.auth0.jwt.JWTCreator;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.ByteArrayEntity;
import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
import org.apache.http.impl.nio.client.HttpAsyncClients;
import org.apache.http.message.BasicHeader;
import org.bouncycastle.jce.ECNamedCurveTable;
import org.bouncycastle.jce.interfaces.ECPublicKey;
import org.bouncycastle.jce.spec.ECNamedCurveParameterSpec;
import org.bouncycastle.math.ec.ECPoint;
import org.jose4j.jws.AlgorithmIdentifiers;
import org.jose4j.jws.JsonWebSignature;
import org.jose4j.jwt.JwtClaims;
import org.jose4j.lang.JoseException;

import com.google.crypto.tink.apps.webpush.WebPushHybridEncrypt;
import com.google.crypto.tink.subtle.EllipticCurves;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;

import java.io.IOException;
import java.net.URL;
import java.nio.charset.Charset;
import java.security.*;
import java.security.interfaces.ECPrivateKey;
import java.security.spec.InvalidKeySpecException;
import java.util.Base64;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

import static nl.martijndwars.webpush.Utils.CURVE;

public class PushService {
    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    /**
     * Subject used in the JWT payload (for VAPID)
     */
    private String subject;

    /**
     * The public key (for VAPID)
     */
    private PublicKey publicKey;

    private String privateKeyString;

    /**
     * The private key (for VAPID)
     */
    private PrivateKey privateKey;

    public PushService() {
    }

    public PushService(String publicKey, String privateKey, String subject) throws GeneralSecurityException {
        this.publicKey = Utils.loadPublicKey(publicKey);
        this.privateKey = Utils.loadPrivateKey(privateKey);
        this.privateKeyString = privateKey;
        this.subject = subject;
    }

    public String vapid_t(Notification message) {
        try {
            URL url = new URL(message.getEndpoint());
            String aud = url.getProtocol() + "://" + url.getHost();

            Date exp = new Date(System.currentTimeMillis() + 12 * 60 * 60 * 1000);

            Algorithm alg = Algorithm.ECDSA256(null, EllipticCurves.getEcPrivateKey(EllipticCurves.CurveType.NIST_P256, privateKeyString.getBytes()));

            return JWT.create().withAudience(aud).withExpiresAt(exp).withSubject(subject).sign(alg);
        } catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }


    public byte[] encrypt(Notification message) {
        try {
            byte[] pk = Utils.savePublicKey((ECPublicKey) publicKey);
            WebPushHybridEncrypt aes128gcm = new WebPushHybridEncrypt.Builder()
                    .withAuthSecret(message.getUserAuth())
                    .withRecipientPublicKey(pk)
                    .build();
            return aes128gcm.encrypt(message.getPayload(), null);
        } catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Send a notification and wait for the response.
     *
     * @param notification
     * @return
     * @throws GeneralSecurityException
     * @throws IOException
     * @throws JoseException
     * @throws ExecutionException
     * @throws InterruptedException
     */
    public HttpResponse send(Notification notification) throws GeneralSecurityException, IOException, JoseException, ExecutionException, InterruptedException {
        return sendAsync(notification).get();
    }

    /**
     * Send a notification, but don't wait for the response.
     *
     * @param notification
     * @return
     * @throws GeneralSecurityException
     * @throws IOException
     * @throws JoseException
     */
    public Future<HttpResponse> sendAsync(Notification notification) throws GeneralSecurityException, IOException, JoseException {
        HttpPost httpPost = preparePost(notification);

        final CloseableHttpAsyncClient closeableHttpAsyncClient = HttpAsyncClients.createSystem();
        closeableHttpAsyncClient.start();

        return closeableHttpAsyncClient.execute(httpPost, new ClosableCallback(closeableHttpAsyncClient));
    }


    /**
     * Prepare a HttpPost for Apache async http client
     *
     * @param notification
     * @return
     * @throws GeneralSecurityException
     * @throws IOException
     * @throws JoseException
     */
    public HttpPost preparePost(Notification notification) throws GeneralSecurityException, IOException, JoseException {
        assert (verifyKeyPair());

        byte[] ciphertext = encrypt(notification);
        String vapid__t = vapid_t(notification);

        HttpPost httpPost = new HttpPost(notification.getEndpoint());
        httpPost.addHeader("TTL", String.valueOf(notification.getTTL()));

        Map<String, String> headers = new HashMap<>();

        if (notification.hasPayload()) {
            headers.put("Content-Type", "application/octet-stream");
            headers.put("Content-Encoding", "aes128gcm");
            httpPost.setEntity(new ByteArrayEntity(ciphertext));
        }

        headers.put("Authorization", "vapid t=" + vapid__t + ",k=" + publicKey); // TODO

        for (Map.Entry<String, String> entry : headers.entrySet()) {
            httpPost.addHeader(new BasicHeader(entry.getKey(), entry.getValue()));
        }
        return httpPost;
    }

    private boolean verifyKeyPair() {
        ECNamedCurveParameterSpec curveParameters = ECNamedCurveTable.getParameterSpec(CURVE);
        ECPoint g = curveParameters.getG();
        ECPoint sG = g.multiply(((ECPrivateKey) privateKey).getS());

        return sG.equals(((ECPublicKey) publicKey).getQ());
    }

    public PublicKey getPublicKey() {
        return publicKey;
    }

    public PrivateKey getPrivateKey() {
        return privateKey;
    }

    public KeyPair getKeyPair() {
        return new KeyPair(publicKey, privateKey);
    }

    /**
     * Check if VAPID is enabled
     *
     * @return
     */
    protected boolean vapidEnabled() {
        return publicKey != null && privateKey != null;
    }
}

File: src/main/java/nl/martijndwars/webpush/PushService.scala
package nl.martijndwars.webpush

import com.google.common.io.BaseEncoding
import org.apache.http.HttpResponse
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.ByteArrayEntity
import org.apache.http.impl.client.HttpClients
import org.apache.http.message.BasicHeader
import org.bouncycastle.jce.ECNamedCurveTable
import org.bouncycastle.jce.interfaces.ECPublicKey
import org.jose4j.jws.AlgorithmIdentifiers
import org.jose4j.jws.JsonWebSignature
import org.jose4j.jwt.JwtClaims
import java.security._

object PushService {
  /**
    * Encrypt the payload using the user's public key using Elliptic Curve
    * Diffie Hellman cryptography over the prime256v1 curve.
    *
    * @return An Encrypted object containing the public key, salt, and
    *         ciphertext, which can be sent to the other party.
    */
  def encrypt(buffer: Array[Byte], userPublicKey: PublicKey, userAuth: Array[Byte]): Encrypted = {
    val parameterSpec = ECNamedCurveTable.getParameterSpec("prime256v1")
    val keyPairGenerator = KeyPairGenerator.getInstance("ECDH", "BC")
    keyPairGenerator.initialize(parameterSpec)
    val serverKey = keyPairGenerator.generateKeyPair
    val keys = Map("server-key-id" -> serverKey)
    val labels = Map("server-key-id" -> "P-256")
    val salt = SecureRandom.getSeed(16)
    val httpEce = new HttpEce(keys, labels)
    val ciphertext = httpEce.encrypt(buffer, salt, null, "server-key-id", userPublicKey, userAuth)
    new Encrypted(publicKey = serverKey.getPublic, salt = salt, ciphertext = ciphertext)
  }
}

final class PushService {

  /**
    * Send a notification
    */
  def send(notification: Notification, publicKey: PublicKey, privateKey: Key, subject: Option[String] = None): HttpResponse = {
    val base64url = BaseEncoding.base64Url
    val encrypted = PushService.encrypt(
      buffer = notification.payload,
      userPublicKey =  notification.userPublicKey,
      userAuth = notification.userAuth
    )
    val dh = Utils.savePublicKey(encrypted.publicKey.asInstanceOf[ECPublicKey])
    val salt = encrypted.salt.value
    val httpPost = new HttpPost(notification.endpoint)
    httpPost.addHeader("TTL", String.valueOf(notification.ttl))
    val headers = new java.util.HashMap[String, String]
    if (notification.hasPayload) {
      headers.put("Content-Type", "application/octet-stream")
      headers.put("Content-Encoding", "aesgcm")
      headers.put("Encryption", "keyid=p256dh;salt=" + base64url.omitPadding.encode(salt))
      headers.put("Crypto-Key", "keyid=p256dh;dh=" + base64url.encode(dh))
      httpPost.setEntity(new ByteArrayEntity(encrypted.ciphertext.value))
    }

    val claims = new JwtClaims
    claims.setAudience(notification.getOrigin)
    claims.setExpirationTimeMinutesInTheFuture(12 * 60)
    claims.setSubject(subject.orNull)
    val jws = new JsonWebSignature
    jws.setHeader("typ", "JWT")
    jws.setHeader("alg", "ES256")
    jws.setPayload(claims.toJson)
    jws.setKey(privateKey)
    jws.setAlgorithmHeaderValue(AlgorithmIdentifiers.ECDSA_USING_P256_CURVE_AND_SHA256)
    headers.put("Authorization", "Bearer " + jws.getCompactSerialization)
    val pk = Utils.savePublicKey(publicKey.asInstanceOf[ECPublicKey])
    if (headers.containsKey("Crypto-Key")) {
      headers.put("Crypto-Key", headers.get("Crypto-Key") + ";p256ecdsa=" + base64url.omitPadding.encode(pk))
    }
    else {
      headers.put("Crypto-Key", "p256ecdsa=" + base64url.encode(pk))
    }

    import scala.collection.JavaConverters._
    for (entry <- headers.entrySet.asScala) {
      httpPost.addHeader(new BasicHeader(entry.getKey, entry.getValue))
    }
    val httpClient = HttpClients.createDefault
    httpClient.execute(httpPost)
  }

}
